### PR TITLE
[C+B] Add Wither API. Adds BUKKIT-4571

### PR DESCRIPTION
**Pull request by t00thpick1**
**Issue:**
There is currently little to nothing that allows developers to alter Wither behavior and such.

**PR breakdown:**
This commit adds methods that would allow developers to control wither behavior whether it be shooting wither heads at certain locations or things, or just stopping them altogether.

**Justification:**
Currently with Bukkit there is no little to allow developers to write code to interface with Wither boss mobs. This means that developers are unable to specify (or retrieve) targets for the Wither on a per-head basis. Furthermore, there is no event or means to detect when a Wither selects a new target for any one of its three heads. This pull request also adds methods to enable developers to fire skulls from one of the three available heads to a location or LivingEntity, something which previously was not possible. This opens up a large number of possibilities for developers and enables finer control over Wither boss mobs than was previously possible.

**CraftBukkit PR:**  https://github.com/Bukkit/CraftBukkit/pull/1208

### CODE REVIEW

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    public boolean isCancelled() {
+        return cancel;
+    }
+
+    public void setCancelled(boolean cancel) {
+        this.cancel = cancel;
+    }
+
+    /**
+     * Get the entity that this is targeting.
```
Comment (lol768) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: This is finished with a full stop whereas others aren't.

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+     *
+     * @return The entity
+     */
+    public LivingEntity getTarget() {
+        return target;
+    }
+
+    /**
+     * Set the entity that you want the mob to target instead.
+     * It is possible to be null, null will cause the entity to be
```
Comment (lol768) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: Maybe reword "Supplying null will cause the entity to be target-less and is ..."

**Comment by GalRedJelly** [comment]
What about an event shooting at a location?

File: src/main/java/org/bukkit/event/entity/EntityTargetEvent.java [hunk 3]
```diff
@@ -116,6 +116,10 @@ public static HandlerList getHandlerList() {
          */
         DEFEND_VILLAGE,
         /**
+         * When an wither selects it's chase target as an attack target for it's middle head
```
Comment (lol768) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java]: `it's` should be `its`

I'd also suggest rewording it, it just sounds a little clunky.

**Comment by t00thpick1** [comment]
@GalRedJelly I plan on doing that within a separate PR, as that will also cover ghasts shooting fireballs and such

File: src/main/java/org/bukkit/event/entity/EntityTargetEvent.java (hunk 3 above)
Comment (t00thpick1) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java]: @lol768 Any ideas on how to word it?

Comment (frymaster) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java]: as a minor nitpick, both your "it's" should be "its"

Comment (lol768) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java]: How about "When a wither boss selects an attack target". You could add "for its middle head" but that's implied by the type of mob.

File: src/main/java/org/bukkit/entity/Wither.java [hunk 4]
```diff
@@ -1,7 +1,46 @@
...
  */
 public interface Wither extends Monster {
+    /**
+     * Gets the current target of a wither head.
+     *
+     * @param head which head to check the target of
+     * @return the current target of a wither head, or null if none
+     */
+    public LivingEntity getTarget(WitherHead head);
+    /**
```
Comment (riking) [inline src/main/java/org/bukkit/entity/Wither.java:18]: There's supposed to be a blank line between the method and the start of the next javadoc
<3

Comment (t00thpick1) [inline src/main/java/org/bukkit/entity/Wither.java:18]: Right you are, fixed

**Comment by turt2live** [comment]
Justification? Testing materials? 
[CONTRIBUTING.md](https://github.com/Bukkit/Bukkit/blob/master/CONTRIBUTING.md) - Please and thank you :)

File: src/main/java/org/bukkit/entity/Wither.java [hunk 5]
```diff
@@ -1,7 +1,51 @@
...
+
+    /**
+     * Fires a wither skull from a wither head.
+     *
+     * @param head which head to fire from
+     * @param location the location to fire at
+     */
+    public void shoot(WitherHead head, Location location);
+
+
```
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Wither.java]: Extra line

File: src/main/java/org/bukkit/entity/Wither.java [hunk 6]
```diff
@@ -1,7 +1,51 @@
...
+     * @param head which head to fire from
+     * @param location the location to fire at
+     */
+    public void shoot(WitherHead head, Location location);
+
+    
+    /**
+     * An enum to specify which head of the wither
+     */
+    public enum WitherHead {
```
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Wither.java:45]: The 3 enum values need JavaDocs

Edit (added example):

``` java
/**
 * Existing JD
 */
public enum WitherHead{

    /**
     * What this does
     */
    LEFT,
    // ETC...
}
```

Note the blank line between `public enum WitherHead{` and the JD for `LEFT`

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+import org.bukkit.entity.Entity;
+import org.bukkit.entity.LivingEntity;
+import org.bukkit.entity.Wither.WitherHead;
+import org.bukkit.event.HandlerList;
+
+/**
+ * Called when a wither head targets or untargets an entity for its wither skull attack
+ */
+public class WitherHeadTargetEvent extends EntityTargetLivingEntityEvent {
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: Could this not be replaced by EntityTargetLivingEntityEvent with the skull head as the "entity" and the target as, well, the target?

Comment (t00thpick1) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: A wither "head" and a wither skull are different objects, each wither "head" represents one of the three visible heads on the wither, which track different targets per each one.  This event is necessary in order to determine which wither head is actually doing the targeting,

File: src/main/java/org/bukkit/entity/Wither.java (hunk 6 above)
Comment (t00thpick1) [inline src/main/java/org/bukkit/entity/Wither.java:45]: I figured Left Right and Center were pretty straightforward but okay.

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java (hunk 7 above)
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: I meant the head, not the skull (oops). 

But fair enough

File: src/main/java/org/bukkit/entity/Wither.java (hunk 6 above)
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Wither.java:45]: So is RIGHT_CLICK_AIR though :/ Everything needs to be documented for sanity

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java (hunk 7 above)
Comment (t00thpick1) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: The heads in themselves are not Entities by themselves, the wither as a whole is one entity, which is why the event was needed in order to determine which of the heads was being used.

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: That is... weird...

And I think you've mentioned this to me before - I just forgot it.

But yes, a new event would make sense for this seeing as the heads are not "entities"

File: src/main/java/org/bukkit/event/entity/EntityTargetEvent.java [hunk 8]
```diff
@@ -116,6 +116,10 @@ public static HandlerList getHandlerList() {
          */
         DEFEND_VILLAGE,
         /**
+         * When a wither boss selects an attack target for its primary head.
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java:125]: `it's`

Comment (t00thpick1) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java:125]: No...  not at all..   "When a wither boss selects an attack target for it is primary head."  or "When a wither boss selects an attack target for its primary head."

Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java:125]: I'll just stop commenting for a while and actually think...

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java (hunk 7 above)
Comment (riking) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: Note that the "skulls" are the projectiles that get shot out, and their lifespans cannot be compared at all to that of the main Wither. For a structured analogy, `[arrows -> players] = [wither skulls -> the wither]`.

File: src/main/java/org/bukkit/entity/Wither.java [hunk 9]
```diff
@@ -1,7 +1,62 @@
...
+    /**
+     * Fires a wither skull from a wither head.
+     *
+     * @param head which head to fire from
+     * @param location the location to fire at
+     */
+    public void shoot(WitherHead head, Location location);
+
+    /**
+     * An enum to specify which head of the wither
```
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Wither.java]: This kinda sounds like an incomplete sentence. Maybe something like `An enum to specify a head on the Wither`?

File: src/main/java/org/bukkit/event/entity/EntityTargetEvent.java [hunk 10]
```diff
@@ -116,6 +116,10 @@ public static HandlerList getHandlerList() {
...
+         */
+        WITHER_TARGET,
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java:127]: Technically there is supposed the be a new line below this and above the JD

```
DEFEND_VILLAGE,

/**
 */
WITHER_TARGET,

/**
```

Edit: However in this case we'll keep it as is as the rest of the file is not correctly spaced

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    private static final HandlerList handlers = new HandlerList();
+    private WitherHead head;
+
+    public WitherHeadTargetEvent(Entity entity, LivingEntity target, TargetReason reason, WitherHead head) {
+        super(entity, target, reason);
+        this.head = head;
+    }
+
+    /**
+     * Get the enum representation of which head is targeting an entity
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: `Gets which head of the wither is targeting an entity`

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+     *
+     * @return Which wither head
```
Comment (turt2live) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: `@return Which wither head is targeting an entity`

Yes, the redundancy is needed

File: src/main/java/org/bukkit/entity/Wither.java [hunk 13]
```diff
@@ -1,7 +1,62 @@
...
+
+        /**
+         * The Center head in respect to a frontal view
+         */
+        CENTER,
+
+        /**
+         * The Right head in respect to a frontal view
+         */
+        RIGHT,
```
Comment (turt2live) [inline src/main/java/org/bukkit/entity/Wither.java]: Shouldn't have a comma after it

**Comment by turt2live** [comment]
Looks good (for now). Is it possible to get a supplied justification and a test plugin (with source) to use?

**Comment by lol768** [comment]
Something along the lines of this perhaps?

**Justification:**

Currently with Bukkit there is no little to allow developers to write code to interface with Wither boss mobs. This means that developers are unable to specify (or retrieve) targets for the Wither on a per-head basis. Furthermore, there is no event or means to detect when a Wither selects a new target for any one of its three heads. This pull request also adds methods to enable developers to fire skulls from one of the three available heads to a location or LivingEntity, something which previously was not possible. This opens up a large number of possibilities for developers and enables finer control over Wither boss mobs than was previously possible.

**Comment by turt2live** [comment]
Perfect, now that needs to end up in the description :3

**Comment by lol768** [comment]
Looks like it's been done.

File: src/main/java/org/bukkit/event/entity/EntityTargetEvent.java (hunk 8 above)
Comment (JeromSar) [inline src/main/java/org/bukkit/event/entity/EntityTargetEvent.java:125]: "it's" is actually a shorthand for "it is". @t00thpick1 clearly means the possessive version of "it".

http://www.thefreedictionary.com/its

File: src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java (hunk 7 above)
Comment (TomyLobo) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: getHeadLocation is missing, returning head position and approximate facing

I think this is also a good time to again request copying Entity.getNearbyEntities and LivingEntity.getLineOfSight/getTargetBlock to the World interface, so we can use it with wither heads.

Comment (t00thpick1) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: Entity.getNearbyEntities() is accessible within the Wither entity itself, and since the head is part of that entity, the method serves for all purposes in regards to withers

Comment (TomyLobo) [inline src/main/java/org/bukkit/event/entity/WitherHeadTargetEvent.java]: the difference is negligible here, yeah, but think about other cases.
Example: find the entities closest to a command block

EDIT: Wither are still a valid use-case for getLineOfSight/getTargetBlock.

**Comment by turt2live** [comment]
Currently this PR is missing the following:
- Testing materials / Proof of concept (just to show that this indeed works)
- The ability to be pulled (a likely 1.7 update may be required)

If anyone can tackle these tasks, the chance of this PR getting pulled significantly increases.

**Comment by t00thpick1** [comment]
As soon as the 1.7 update hits the repo I will work on getting it synced up.  No worries on that.

**Comment by TomyLobo** [comment]
Can you get these into CraftWitherSkull, too?

``` java
@Override
public void setCharged(boolean charged) {
    getHandle().a(charged);
}

@Override
public boolean getCharged() {
    return getHandle().d();
}
```

plus the bukkit side of things of course :)

**Comment by riking** [comment]
And what's the definition of charged? That it's flashing white and only
accepts melee damage?

Any chance of vanilla overwriting it?

**Comment by TomyLobo** [comment]
i dunno, someone in #bukkitdev wanted to make wither skulls glow blue, so i came up with this.
suggest a better name if you have one.

(I actually never ran that code, but these seems to be the only ones anyway, so i guess they must correspond to the blue glow)

EDIT: I re-read your post. no those arent for the wither, but for the wither skull.

**Comment by t00thpick1** [comment]
@TomyLobo Once 1.7 hits I plan on doing a larger project to get entities that fire Projectiles such as withers shooting heads and ghasts shooting fireballs an event or two,  that PR will include the ability to edit wither skull charged status

**Comment by TomyLobo** [comment]
neat, don't forget to make "hit" events for projectiles.
Potions could use a subclass of that event with an additional "affected" property containing the ents in its radius.

**Comment by t00thpick1** [comment]
Rebased to 1.7,  implementation also patched up for 1.7

**Comment by turt2live** [comment]
@t00thpick1 If you are still interested in this, please update it to the latest code base.

Thanks!

**Comment by t00thpick1** [comment]
Rebased to current.

**Comment by turt2live** [comment]
@t00thpick1 If you, or anyone else, is still interested in this, please address the following concerns:
- The ability to be pulled (update to latest master)
- Testing materials (3rd request)

Thanks!

**Comment by TomyLobo** [comment]
I think in the long run, unit tests would be more useful than a test plugin.

**Comment by turt2live** [comment]
@TomyLobo we still need to test it in-game, so a test plugin is still required for this. Unit tests, although useful, may be requested after the actual code has been tested to ensure it's at the functionality we want.

**Comment by keir-nellyer** [comment]
cc https://github.com/Bukkit/CraftBukkit/pull/1208#issuecomment-49558492
